### PR TITLE
Increased use of NodeIndexSeek to wider range of cases

### PR DESCRIPTION
**Pull request by eebus**
changelog: Increased use of NodeIndexSeek to wider range of cases.

* Equality predicates on expressions of map literals and variables of map type
* Index Nested Loop joins even when incoming horizon contained a query plan

### CODE REVIEW

**Comment by eebus** [comment]
[GH# 8463 USING INDEX causes error even though property value comparisons are now supported](https://trello.com/c/BGwoyjGi/53-gh-8463-using-index-causes-error-even-though-property-value-comparisons-are-now-supported)

**Comment by eebus** [comment]
[Allow NodeIndexSeek on predicates that involve other node properties](https://trello.com/c/CDVPYF3Q/58-allow-nodeindexseek-on-predicates-that-involve-other-node-properties)

File: community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexNestedLoopJoinAcceptanceTest.scala [hunk 1]
```diff
@@ -72,4 +73,107 @@ class IndexNestedLoopJoinAcceptanceTest extends ExecutionEngineFunSuite with New
     result should use("Apply", "NodeIndexSeek")
     result should not(use("ValueHashJoin", "CartesianProduct", "NodeByLabelScan", "Filter"))
   }
+
+  test("should use index on literal value") {
```
Comment (systay) [inline community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexNestedLoopJoinAcceptanceTest.scala]: This test and the next two tests has nothing to do with NestedIndexLoopJoins

File: community/cypher/frontend-3.1/src/main/scala/org/neo4j/cypher/internal/frontend/v3_1/ast/Clause.scala [hunk 2]
```diff
@@ -180,7 +180,7 @@ case class Match(optional: Boolean, pattern: Pattern, hints: Seq[UsingHint], whe
    */
   private def applicable(other: Expression) = {
     other match {
-      case _: Property => false
+//      case _: Property => false
```
Comment (systay) [inline community/cypher/frontend-3.1/src/main/scala/org/neo4j/cypher/internal/frontend/v3_1/ast/Clause.scala]: This line should be removed, right?

**Review by systay — CHANGES_REQUESTED** [review]

**Comment by eebus** [comment]
[[3.0.3] No index hints possible with values from LOAD CSV row](https://trello.com/c/Y0tov1LD/70-3-0-3-no-index-hints-possible-with-values-from-load-csv-row)

File: community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexNestedLoopJoinAcceptanceTest.scala [hunk 3]
```diff
@@ -72,4 +73,74 @@ class IndexNestedLoopJoinAcceptanceTest extends ExecutionEngineFunSuite with New
...
+
+  test("should use index on literal map expression") {
+    val nodes = Range(0,125).map(i => createLabeledNode(Map("id" -> i), "Foo"))
+    graph.createIndex("Foo", "id")
+    val query =
+      """
+        |PROFILE
+        | MATCH (f:Foo)
+        | USING INDEX f:Foo(id)
+        | WHERE f.id={id: 123}.id
```
Comment (Mats-SX) [inline community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexNestedLoopJoinAcceptanceTest.scala]: Hah! Nice.

File: community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/SemanticErrorAcceptanceTest.scala [hunk 4]
```diff
@@ -282,15 +282,15 @@ class SemanticErrorAcceptanceTest extends ExecutionEngineFunSuite {
     )
   }
 
-  test("should fail if using a hint with property equality comparison") {
+  test("should fail if using a hint with multiple OR'd equality comparisons") {
     executeAndEnsureError(
-      "match (n:Person)-->(m:Person) using index n:Person(name) where n.name = m.name return n",
+      "match (n:Person) using index n:Person(name) where n.name = 'a' OR n.name = 'b' return n",
```
Comment (Mats-SX) [inline community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/SemanticErrorAcceptanceTest.scala:287]: Why was this changed? Are we now supporting queries on the form that was replaced here?

File: community/cypher/frontend-3.1/src/main/scala/org/neo4j/cypher/internal/frontend/v3_1/ast/Clause.scala [hunk 5]
```diff
@@ -180,7 +180,6 @@ case class Match(optional: Boolean, pattern: Pattern, hints: Seq[UsingHint], whe
    */
   private def applicable(other: Expression) = {
     other match {
-      case _: Property => false
```
Comment (Mats-SX) [inline community/cypher/frontend-3.1/src/main/scala/org/neo4j/cypher/internal/frontend/v3_1/ast/Clause.scala:183]: Is this the only actual change??

Comment (eebus) [reply community/cypher/frontend-3.1/src/main/scala/org/neo4j/cypher/internal/frontend/v3_1/ast/Clause.scala:183]: This and the change in the file cartesianProductsOrValueJoins.scala

File: community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/SemanticErrorAcceptanceTest.scala (hunk 4 above)
Comment (craigtaverner) [reply community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/SemanticErrorAcceptanceTest.scala:287]: Correct. We support 'Nested Index Loop Joins' for this case. But the error can be generated for other cases, so this test for error message needed a different query to generate that error.

**Review by systay — APPROVED** [review]

File: community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexNestedLoopJoinAcceptanceTest.scala (hunk 3 above)
Comment (systay) [reply community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexNestedLoopJoinAcceptanceTest.scala]: Nice test, but does it really belong here? This is not handled by the NIJ component, right?

File: community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/LoadCsvAcceptanceTest.scala [hunk 6]
```diff
@@ -43,6 +43,40 @@ class LoadCsvAcceptanceTest
...
+        s"""LOAD CSV WITH HEADERS FROM '$url' AS row
+            | MATCH (user:User{userID: row.USERID}) USING INDEX user:User(userID)
+            | MATCH (order:Order{orderID: row.OrderId})
+            | MATCH (user)-[acc:ORDERED]->(order)
+            | SET acc.field1=row.field1,
+            | acc.field2=row.field2
+            | RETURN count(*); """.stripMargin
+      )
+
+      assertStats(result, propertiesWritten = 6)
```
Comment (systay) [inline community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/LoadCsvAcceptanceTest.scala:77]: Should we not also test that the index was actually used?

File: community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexNestedLoopJoinAcceptanceTest.scala (hunk 3 above)
Comment (eebus) [reply community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexNestedLoopJoinAcceptanceTest.scala]: Agree

File: community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/LoadCsvAcceptanceTest.scala (hunk 6 above)
Comment (eebus) [reply community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/LoadCsvAcceptanceTest.scala:77]: Yes

**Pull request merged**